src/main/java/org/tugraz/sysds/runtime/util/DataConverter.java

```diff
@@ -904,6 +908,149 @@ public static String toString(MatrixBlock mb, boolean sparse, String separator,
 		return sb.toString();
 	}
 
+	public static String toString(TensorBlock mb) {
+		return toString(mb, false, " ", "\n", "[", "]", mb.getNumRows(), mb.getNumCols(), 3);
+	}
+
+	/**
+	 * Returns a string representation of a tensor
+	 * @param tb tensor block
+	 * @param sparse if true, string will contain a table with row index, col index, value (where value != 0.0)
+	 * 				 otherwise it will be a rectangular string with all values of the tensor block
+	 * @param separator Separator string between each element in a row, or between the columns in sparse format
+	 * @param lineseparator Separator string between each row
+	 * @param leftBorder Characters placed at the start of a new dimension level
+	 * @param rightBorder Characters placed at the end of a new dimension level
+	 * @param rowsToPrint maximum number of rows to print, -1 for all
+	 * @param colsToPrint maximum number of columns to print, -1 for all
+	 * @param decimal number of decimal places to print, -1 for default
+	 * @return tensor as a string
+	 */
+	public static String toString(TensorBlock tb, boolean sparse, String separator, String lineseparator,
+	                              String leftBorder, String rightBorder, int rowsToPrint, int colsToPrint, int decimal){
+		StringBuilder sb = new StringBuilder();
+
+		// Setup number of rows and columns to print
+		int rlen = tb.getNumRows();
+		int clen = tb.getNumCols();
+		int rowLength = rlen;
+		int colLength = clen;
+		if (rowsToPrint >= 0)
+			rowLength = Math.min(rowsToPrint, rlen);
+		if (colsToPrint >= 0)
+			colLength = Math.min(colsToPrint, clen);
+
+		DecimalFormat df = new DecimalFormat();
+		df.setGroupingUsed(false);
+		if (decimal >= 0){
+			df.setMinimumFractionDigits(decimal);
+		}
+
+		if (sparse){ // Sparse Print Format
+			// TODO use sparse iterator for sparse block
+			int[] ix = new int[tb.getNumDims()];
+			for (int i = 0; i < tb.getLength(); i++) {
+				int j = ix.length - 1;
+				String str = tb.getString(ix);
+				if (str != null && !str.isEmpty() && Double.parseDouble(str) != 0) {
+					for (int item : ix) {
+						sb.append(item).append(separator);
+					}
+					concatenateTensorValue(tb, sb, df, ix);
+					sb.append(lineseparator);
+				}
+				ix[j]++;
+				//calculating next index
+				if (ix[j] == tb.getDim(j)) {
+					while (ix[j] == tb.getDim(j) || ix[1] >= colLength) {
```

Comment:
I've seen a similar code fragment before, so please either (1) simplify to some easier-to-understand for loop, or (2) encapsulate it in a function for index computation.

Reply:
I encapsulated the loop for the sparse print format, for dense though using the encapsulated method leads to even harder to understand code so I simplified it a bit and added comments. I hope it is now clear enough.